File: python/get_primes.py
import math
import sys
import time

start = time.time()
primes = []
n_primes_target = int(sys.argv[1])
n_primes_found = len(primes)

numerator = 2
denominator = 2
limit = 0

print(f'primes_target = {n_primes_target}')

while n_primes_found < n_primes_target:
    limit = int(math.sqrt(numerator))
    for denominator in range(2, (limit + 2)):
        if (numerator % denominator) == 0:
            break

    # Got to here by one of two ways.
    # If here by break, then numerator is not a prime.
    # If denominator exceeds square root of numerator, 
    # then numerator is a prime.
    if denominator > limit:
        primes.append(numerator)
        n_primes_found = len(primes)

    numerator += 1

end = time.time()
sum = 0
for prime in primes:
    sum += prime
    # print(prime)

duration = int(1000 * (end - start))
print(f'Sum = {sum} of first {n_primes_found} prime numbers.')
print(f'Duration: {duration} [ms] (in Python)') 

File: c/get_primes.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>

/*
Compile with gcc -Wall -O3 get_primes.c -o get_primes.exe
*/

int main(int argc, char *argv[]) {
    struct timeval start, end;
    int *pMem = NULL;
    int limit = 0;
    int nPrimesFound = 0;
    unsigned long sum = 0; 
    int microSecs = 0; 
    int numerator = 5;
    int denominator = 2;
    int nPrimesTarget = atoi(argv[1]);
    
    gettimeofday(&start, NULL);
    
    printf("Getting first %d prime numbers\n", nPrimesTarget);
    pMem = (int*) malloc(nPrimesTarget * sizeof(int));
    pMem[nPrimesFound++] = 2;
    pMem[nPrimesFound++] = 3;

    while (nPrimesFound < nPrimesTarget) {
        limit = (int) sqrt(numerator);
        for (denominator = 2; denominator <= limit; denominator++) {
            if ((numerator % denominator) == 0) {
                break;
            }
        }

        /* Got to here by one of two ways:
         * If by break then numerator is not a prime.
         * If denominator is greater that the limit, then numerator is prime.
        */
        if (denominator > limit) {
            pMem[nPrimesFound] = numerator;
            nPrimesFound++;
        }
        
        numerator += 2;
    }

    sum = 0;
    for (int i = 0; i < nPrimesFound; i++) {
        sum += pMem[i];
        /* 
        printf("%d\n", pMem[i]);
        */
    }

    free(pMem); 
    gettimeofday(&end, NULL);

    microSecs = (1000000 * (end.tv_sec - start.tv_sec) + end.tv_usec - start.tv_usec);
    printf("Sum = %lu of first %d prime numbers.\n", sum, nPrimesFound);
    printf("Duration:  %d [ms] (in C) \n", microSecs / 1000);
    
    return 0;
}    


